source-firestore/pull.go

```diff
@@ -0,0 +1,322 @@
+			} else if msg.Acknowledge == nil {
+				return fmt.Errorf("expected PullRequest.Acknowledge, got %#v", msg)
+			}
+			select {
+			case <-ctx.Done():
+				return nil
+			case acks <- msg.Acknowledge:
+				// Loop and receive another acknowledgement
+			}
+		}
+	})
+	c.Acks = acks
+
+	// Notify Flow that we're starting.
+	log.Debug("sending PullResponse.Opened")
+	if err := c.Stream.Send(&pc.PullResponse{Opened: &pc.PullResponse_Opened{}}); err != nil {
+		return fmt.Errorf("error sending PullResponse.Opened: %w", err)
+	}
+
+	eg.Go(func() error {
+		return c.Capture(ctx)
+	})
+	return eg.Wait()
+}
+
+func (c *capture) Capture(ctx context.Context) error {
+	var scanInterval, _ = time.ParseDuration(DEFAULT_SCAN_INTERVAL)
+	if c.Config.ScanInterval != "" && c.Config.ScanInterval != SCAN_INTERVAL_NEVER {
+		var err error
+		scanInterval, err = time.ParseDuration(c.Config.ScanInterval)
+
+		if err != nil {
+			return fmt.Errorf("parsing scan interval failed: %w", err)
+		}
+	}
+
+	// Listening on snapshots gives you a copy of all documents initially
+	// so in essence it is a full scan. Hence we consider the current moment to be the last scan, and then start from there
+	var lastScan = time.Now()
+
+	sa := option.WithCredentialsJSON([]byte(c.Config.CredentialsJSON))
+	app, err := firebase.NewApp(ctx, nil, sa)
+	if err != nil {
+		return err
+	}
+
+	client, err := app.Firestore(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	var eg = new(errgroup.Group)
+
+	for _, binding := range c.Bindings {
+		var groupID = getLastCollectionGroupID(binding.Path)
+		var query firestore.Query
+
+		if isRootCollection(binding.Path) {
+			query = client.Collection(binding.Path).Query
+		} else {
+			query = client.CollectionGroup(getLastCollectionGroupID(binding.Path)).Query
+		}
+
+		if c.Config.ScanInterval != SCAN_INTERVAL_NEVER {
+			log.WithField("group", groupID).Debug("full scan")
+			eg.Go(func() error {
+				return c.FullScan(ctx, lastScan, scanInterval, query)
+			})
+		}
+
+		if c.Tail {
+			log.WithField("group", groupID).Debug("listening for changes")
+			eg.Go(func() error {
+				return c.ListenForChanges(ctx, query)
+			})
+		}
+	}
+
+	eg.Go(func() error {
+		return c.EmitDocuments()
+	})
+
+	if err := eg.Wait(); err != nil {
+		return err
+	}
+
+	close(c.Output)
+
+	return nil
+}
+
+// TODO: listen on all changes on all documents
+func (c *capture) ListenForChanges(ctx context.Context, query firestore.Query) error {
+	var snapshotsIterator = query.Snapshots(ctx)
+
+	for {
+		var snapshot, err = snapshotsIterator.Next()
+
+		if err == iterator.Done {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		for _, change := range snapshot.Changes {
+			if change.Kind == firestore.DocumentAdded || change.Kind == firestore.DocumentModified {
+
+				// Make sure this document matches the binding path
+				var index = findBindingForCollectionPath(change.Doc.Ref.Parent.Path, c.Bindings)
```

Comment:
is it possible that it wouldn't? I suppose an example would be nested collections not being captured?

Related question: to capture a nested collection, you MUST capture it's parent?

Reply:
@jgraettinger yes, below I check for `index == -1` which is the case where the document doesn't match any of the bindings, in which case we just skip this document.

I haven't tested this, but no, you should not need to capture a parent collection to capture a nested collection thanks to [CollectionGroup](https://firebase.googleblog.com/2019/06/understanding-collection-group-queries.html) queries